### PR TITLE
Merge resumption callbacks and adalr lion

### PR DESCRIPTION
**Pull request by bmosaicml**
This PR adds my two most successful Lion variants ClipLion and AdaLRLion which (resp) clip gradients that are outliers or scale down the LR when the moment becomes an outlier.

I also add my two most successful resumption callbacks GlobalLRLowering and LayerFreezing

image

### CODE REVIEW

File: examples/common/builders.py [hunk 1]
```diff
@@ -11,14 +11,15 @@
 from composer.datasets.in_context_learning_evaluation import \
     get_icl_task_dataloader
 from composer.loggers import WandBLogger
-from composer.optim import DecoupledAdamW
+from composer.optim import DecoupledAdamW, ClipLion, AdaLRLion
```
Comment (vchiley) [inline examples/common/builders.py:14]:
```suggestion
from composer.optim import DecoupledAdamW
```

File: examples/common/resumption_callbacks/__init__.py [hunk 2]
```diff
@@ -0,0 +1,6 @@
+from examples.common.resumption_callbacks.callbacks import LayerFreezing, GlobalLRScaling
+
+RESUMPTION_STRATEGIES = {
+    "layer_freezing": LayerFreezing,
+    'global_lr_scaling': GlobalLRScaling,
+}
```
Comment (vchiley) [inline examples/common/resumption_callbacks/__init__.py:6]:
```suggestion
}

```
new line needed

File: examples/common/optim/outlier_detection.py [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+            slow_mva = sum(self.slow_moving_average) / len(self.slow_moving_average)
+            return obs > self.outlier_threshold * slow_mva
+        else:
+            return False
+    
+    def get_slow_mva(self):
+        if len(self.slow_moving_average) > 0:
+            return sum(self.slow_moving_average) / len(self.slow_moving_average)
+        else:
+            return 0
```
Comment (vchiley) [inline examples/common/optim/outlier_detection.py:32]:
```suggestion
            return 0

```
new line

File: examples/common/builders.py [hunk 4]
```diff
@@ -39,6 +40,9 @@ def build_callback(name, kwargs):
             'log_optimizer_metrics', True),)
     elif name == 'health_checker':
         return HealthChecker(**kwargs)
+    elif name in RESUMPTION_STRATEGIES:
```
Comment (vchiley) [inline examples/common/builders.py:43]:
```suggestion
    elif name in RESUMPTION_STRATEGIES.keys():
```

makes it explicit that this is a dict... idk you can ignore if you want.

File: examples/common/builders.py [hunk 5]
```diff
@@ -77,7 +81,21 @@ def build_optimizer(cfg, model):
...
+                              outlier_threshold=cfg.outlier_threshold)
+    elif cfg.name == 'adalr_lion':
+        return AdaLRLion(model.parameters(),
+                              lr=cfg.lr,
+                              betas=cfg.betas,
+                              weight_decay=cfg.weight_decay,
+                              outlier_threshold=cfg.outlier_threshold,
+                              timeout=cfg.timeout,
+                              lr_penalty=cfg.lr_penalty,
+                              min_scale=cfg.min_scale)
```
Comment (vchiley) [inline examples/common/builders.py:98]:
```suggestion
    elif cfg.name == 'clip_lion':
        return ClipLion(model.parameters(),
                        lr=cfg.lr,
                        betas=cfg.betas,
                        weight_decay=cfg.weight_decay,
                        outlier_threshold=cfg.outlier_threshold)
    elif cfg.name == 'adalr_lion':
        return AdaLRLion(model.parameters(),
                         lr=cfg.lr,
                         betas=cfg.betas,
                         weight_decay=cfg.weight_decay,
                         outlier_threshold=cfg.outlier_threshold,
                         timeout=cfg.timeout,
                         lr_penalty=cfg.lr_penalty,
                         min_scale=cfg.min_scale)
```

lint

File: examples/common/resumption_callbacks/callbacks.py [hunk 6]
```diff
@@ -0,0 +1,50 @@
+from composer.core import Callback, State
+from composer.loggers import Logger
+from typing import List
+
+__all__ = ['ResumptionCallback', 'GlobalLRScaling', 'LayerFreezing',]
+
+
+class ResumptionCallback(Callback):
```
Comment (vchiley) [inline examples/common/resumption_callbacks/callbacks.py:8]: do we want to label this as a AbstractBaseClass

File: examples/common/resumption_callbacks/callbacks.py [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+    def __init__(self,):
+          pass
```
Comment (vchiley) [inline examples/common/resumption_callbacks/callbacks.py:10]: Y even define `def __init__(self,): ` if all we do is `pass`?
Y not just:
```
class ResumptionCallback(Callback):
    pass
```

File: examples/common/resumption_callbacks/callbacks.py (hunk 6 above)
Comment (bmosaicml) [reply examples/common/resumption_callbacks/callbacks.py]: good idea!

File: examples/common/builders.py [hunk 8]
```diff
@@ -39,6 +40,9 @@ def build_callback(name, kwargs):
             'log_optimizer_metrics', True),)
     elif name == 'health_checker':
         return HealthChecker(**kwargs)
+    elif name in RESUMPTION_STRATEGIES.keys():
+        resumption_args = kwargs.get('config', {})
+        return RESUMPTION_STRATEGIES[name](**resumption_args)
```
Comment (abhi-mosaic) [inline examples/common/builders.py:45]: I'm not a fan of adding another layer of registries at this time. Lets add them all here spelled out.

File: examples/common/resumption_callbacks/callbacks.py [hunk 9]
```diff
@@ -0,0 +1,50 @@
+from composer.core import Callback, State
+from composer.loggers import Logger
+from typing import List
+from abc import ABC
+
+__all__ = ['ResumptionCallback', 'GlobalLRScaling', 'LayerFreezing',]
+
+
+class ResumptionCallback(Callback, ABC):
+    pass
```
Comment (abhi-mosaic) [inline examples/common/resumption_callbacks/callbacks.py:10]: @bmosaicml are there any plans for this subclass to have specific traits that a standard `Callback` does not have? If there is no difference, then I would prefer to:

* remove `ResumptionCallback` and make all functions below just subclass `Callback`
* rename this file `resumption_callbacks.py` and move it to `examples/common/resumption_callbacks.py`

**Review by abhi-mosaic — COMMENTED** [review]
Couple cosmetic comments, but otherwise this looks great and I'm glad to have these in our pocket for training stability!

File: examples/common/resumption_callbacks/callbacks.py (hunk 9 above)
Comment (bmosaicml) [reply examples/common/resumption_callbacks/callbacks.py:10]: yeah the sounds good to me!

File: examples/common/optim/outlier_detection.py [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+    
+    def __init__(self,
+                 threshold=7.5,
+                 increase_lookback=500):
+          
+        self.intermediate_data_queue = collections.deque(maxlen=increase_lookback)
+        self.slow_moving_average = collections.deque(maxlen=increase_lookback)
+        self.outlier_threshold = threshold
+    
+    def insert_observation(self, obs):
```
Comment (dakinggg) [inline examples/common/optim/outlier_detection.py:14]: would be good to add a docstring here explaining that this function both inserts the item into the tracking queues, and returns whether the item is considered an outlier. Might also change the function name to reflect this. Or, if reasonable, might be nicer to break up into two functions, one which checks if an observation is an outlier, and one which inserts the observation into the tracking queues.

File: examples/common/optim/adaptive_lion.py [hunk 11]
```diff
@@ -0,0 +1,401 @@
...
+                param.grad.flatten(), optim_state['exp_avg'].flatten(), dim=0),
+    }
+
+    def __init__(
+        self,
+        params,
+        lr: float = 1e-4,
+        betas: Tuple[float, float] = (0.9, 0.99),
+        weight_decay: float = 0.0,
+        outlier_threshold = 10.0,
```
Comment (dakinggg) [inline examples/common/optim/adaptive_lion.py:38]:
```suggestion
        outlier_threshold: float = 10.0,
```

File: examples/common/optim/adaptive_lion.py [hunk 12]
```diff
@@ -0,0 +1,401 @@
...
+                
+                removed = []
+                for ts in state['outlier_timestamp']:
+                    if state['step'] - ts > self.timeout:
+                        removed.append(ts)
+                
+                for ts in removed:
+                    state['outlier_timestamp'].remove(ts)
+                
+                lr = self.adjust_lr(lr, self.lr_penalty, len(state['outlier_timestamp']), self.min_scale)
```
Comment (dakinggg) [inline examples/common/optim/adaptive_lion.py:129]: am i reading this correctly that you basically decrease the learning rate proportional to the number of outliers in the previous `timeout` number of steps? would be good to add an explanation of the algorithm to the docstring

File: examples/common/optim/adaptive_lion.py [hunk 13]
```diff
@@ -0,0 +1,401 @@
...
+            for metric in self.metric_functions:
+                optimizer_metrics[f'{metric}/{name}'] = self.metric_functions[metric](param, param_optim_state,
+                                                                                      step_tensor)
+
+            optimizer_metrics[f'layerwise_lr/{name}'] = torch.tensor(layerwise_lr)
+
+        return optimizer_metrics
+
+
+class ClipLion(Optimizer):
```
Comment (dakinggg) [inline examples/common/optim/adaptive_lion.py:219]: docstring please

File: examples/common/optim/adaptive_lion.py [hunk 14]
```diff
@@ -0,0 +1,401 @@
...
+
+    def __init__(
+        self,
+        params,
+        lr: float = 1e-4,
+        betas: Tuple[float, float] = (0.9, 0.99),
+        weight_decay: float = 0.0,
+        outlier_threshold = 5.0
+    ):
+        assert lr > 0.
```
Comment (dakinggg) [inline examples/common/optim/adaptive_lion.py:245]: raise instead of assert for these

File: examples/common/builders.py [hunk 15]
```diff
@@ -17,8 +17,9 @@
                                       LinearWithWarmupScheduler)
 
 from examples.common.fdiff import FDiffMetrics
-from examples.common.optim.lion import DecoupledLionW
+from examples.common.optim import DecoupledLionW, ClipLion, AdaLRLion
```
Comment (vchiley) [inline examples/common/builders.py:20]: Should we also call these "decoupled"
```suggestion
from examples.common.optim import DecoupledLionW, DecoupledClipLion, DecoupledAdaLRLion
```

**Review by vchiley — COMMENTED** [review]
added a new comments.

Could you also add a few wandb plots showing this is a good idea, or link to a jira, or experimentation log, or something

**Review by vchiley — APPROVED** [review]
looks good.
(lint will prob catch a few things)

If you could add a few plots to the description, that'd be awesome.
Note: I didn't check algos since this is he baseline algo 😄
